src/collector/mod.rs

```diff
@@ -60,6 +61,70 @@ impl From<CollectorOptions> for WebSocketListenerConfig {
 	}
 }
 
+/// This type is used to distinguish different keys in the storage
+#[derive(Clone, Copy, Debug, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub(crate) enum CollectorPrefixType {
+	/// Candidate prefixed by ParachainId
+	Candidate(Option<u32>),
+	/// Relay chain block
+	Head,
+}
+
+/// Used to query storage for multiple hash types
+#[derive(Clone, Debug, Ord, PartialOrd, Eq)]
+pub(crate) struct CollectorKey {
+	/// Prefix for a specific collector key
+	pub prefix: CollectorPrefixType,
+	/// Block or a candidate hash
+	pub hash: H256,
+}
+
+impl CollectorKey {
+	pub(crate) fn new_relay_chain_block(hash: H256) -> Self {
+		Self { prefix: CollectorPrefixType::Head, hash }
+	}
+	pub(crate) fn new_parachain_candidate(hash: H256, para_id: u32) -> Self {
+		Self { prefix: CollectorPrefixType::Candidate(Some(para_id)), hash }
+	}
+	// This is used when we don't know the parachain id
+	pub(crate) fn new_generic_hash(hash: H256) -> Self {
+		Self { prefix: CollectorPrefixType::Candidate(None), hash }
+	}
+
+	pub(crate) fn new_with_prefix(prefix: CollectorPrefixType) -> Self {
+		Self { prefix, hash: H256::zero() }
+	}
+}
+
+impl HasPrefix for CollectorKey {
+	fn has_prefix(&self, other: &Self) -> bool {
```

Comment:
Does this actually check for the same prefix ? It is a bit confusing, the trait definition says:
```
	/// Returns whether there is a specified prefix at the start
	/// The default implementation assumes that there is no prefix at all
```

Reply:
I am not sure we even need this trait if we consider it just as the first part of the key blob. This logic would just be internal to the storage module.

Reply:
Well, I have another thought, that we should actually have a storage that implements `keys_prefix` when the key type T satisfies `HasPrefix` trait. It might be a more convenient interface. With regard to the first comment, I think this should be a part of another PR to fix collector.

Reply:
I have implemented this approach. So the storage is normal by default but can optionally be created with prefixes support when `K` has `HasPrefix` trait. It will leave the remaining tools untouched (and there no longer a requirement to implement `HasPrefix` for untouched types) and allows collector to do prefixed stuff.